### PR TITLE
Allow to use collection for block childrens

### PR DESCRIPTION
**Pull request by VincentLanglet**
## Subject

I am targeting this branch, because {reason}.

Closes #{put_issue_number_here}.

## Changelog

```markdown
### Added
- BlockInterface::addChild()
- BlockInterface::hasChild()

### Changed
- BlockInterface::getChildren() now returns iterable

### Deprecated
- BlockInterface::addChildren()
- BlockInterface::hasChildren()
```

### CODE REVIEW

File: src/Model/BaseBlock.php [hunk 1]
```diff
@@ -44,7 +46,9 @@ abstract class BaseBlock implements BlockInterface
     protected $parent;
 
     /**
-     * @var BlockInterface[]
+     * NEXT_MAJOR: Restrict typehint to Collection.
+     *
+     * @var Collection<int, BlockInterface>|array<BlockInterface>
```
Comment (jordisala1991) [inline src/Model/BaseBlock.php:51]: what is the advantage over iterable?

File: src/Model/BaseBlock.php [hunk 2]
```diff
@@ -170,14 +174,32 @@ public function getUpdatedAt(): ?\DateTime
...
 
+    public function addChild(BlockInterface $child): void
+    {
+        $this->children[] = $child;
+
+        $child->setParent($this);
+    }
+
+    /**
+     * NEXT_MAJOR: Remove this method.
```
Comment (jordisala1991) [inline src/Model/BaseBlock.php:185]: do we use addChild on page bundle?

File: src/Model/BaseBlock.php (hunk 1 above)
Comment (VincentLanglet) [reply src/Model/BaseBlock.php:51]: We're using `count`: https://github.com/sonata-project/SonataBlockBundle/blob/4.x/src/Model/BaseBlock.php#L224

File: src/Model/BaseBlock.php (hunk 2 above)
Comment (VincentLanglet) [reply src/Model/BaseBlock.php:185]: No.

In the same way we have `addChildren` method in the PageInterface of PageBundle https://github.com/sonata-project/SonataPageBundle/blob/fed26c06856de8d6b1161d0f23001db7b8377e2b/src/Model/PageInterface.php#L189 which is not used. We could rename this to `addChild`.

If you had in mind to remove the addChild method. I'm not chocked having some add/set method in the interface/class which will only be used by the user.

File: src/Model/BaseBlock.php (hunk 1 above)
Comment (dmaicher) [reply src/Model/BaseBlock.php:51]: so `array<BlockInterface>` would be fine then? :thinking:

Comment (VincentLanglet) [reply src/Model/BaseBlock.php:51]: The goal is to allow PageBundle to use Collection, https://github.com/sonata-project/SonataPageBundle/blob/4.x/src/Entity/BaseBlock.php#L34

See https://github.com/sonata-project/SonataPageBundle/issues/1442 or https://github.com/sonata-project/SonataPageBundle/pull/1469#discussion_r923635402

**Review by jordisala1991 — APPROVED** [review]

**Review by dmaicher — APPROVED** [review]

**Pull request merged**